File: 프로그래머스/2/42860. 조이스틱/조이스틱.cpp
#include <bits/stdc++.h>
using namespace std;

int A[26];

int solution(string name) {
    int ans = 0;
    int len = name.size();
    int cmp = len - 1;
    
    // 거리 초기화
    for(int i=0; i<=13; i++) A[i] = i;
    for(int j=12, i=14; i<26; i++, j--) A[i] = j;
    
    for(int i=0; i<len; i++) {
        ans += A[name[i] - 'A'];
        
        int tmp = i + 1;
        while(tmp < len && name[tmp] == 'A') tmp++;
        cmp = min(cmp, i + len - tmp + min(i, len - tmp));
    }
    
    return ans + cmp;
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
def solution(name):
    ans = 0
    cmp = len(name) - 1

    for i, ch in enumerate(name):
        ans += min(ord(ch) - ord('A'), ord('Z') - ord(ch) + 1)

        nxt = i + 1
        while nxt < len(name) and name[nxt] == 'A':
            nxt += 1
        
        cmp = min([cmp, 2 * i + len(name) - nxt, i + 2 * (len(name) - nxt)])
    
    return ans + cmp